contracts/upkeeps/CronUpkeep.sol

```diff
@@ -0,0 +1,311 @@
+    external
+    view
+    returns (
+      address target,
+      bytes memory handler,
+      string memory cronString,
+      uint256 nextTick
+    )
+  {
+    Spec memory spec = s_specs[id];
+    Spec_External memory spec_ex = abi.decode(
+      abi.encode(spec),
+      (Spec_External)
+    );
+    return (
+      s_targets[id],
+      s_handlers[id],
+      spec_ex.toCronString(),
+      spec_ex.nextTick()
+    );
+  }
+
+  /**
+   * @notice Converts a cron string to a Spec, validates the spec, and encodes the spec.
+   * This should only be called off-chain, as it is gas expensive!
+   * @param cronString the cron string to convert and encode
+   * @return the abi encoding of the Spec struct representing the cron string
+   */
+  function cronStringToEncodedSpec(string memory cronString)
+    external
+    pure
+    returns (bytes memory)
+  {
+    return cronString.toEncodedSpec();
+  }
+
+  /**
+   * @notice Adds a cron spec to storage and the ID to the list of jobs
+   * @param target the destination contract of a cron job
+   * @param handler the function signature on the target contract to call
+   * @param spec the cron spec to create
+   */
+  function createCronJobFromSpec(
+    address target,
+    bytes memory handler,
+    Spec memory spec
+  ) internal onlyOwner {
+    uint256 newID = s_nextCronJobID;
+    s_activeCronJobIDs.push(newID);
+    s_targets[newID] = target;
+    s_handlers[newID] = handler;
+    s_specs[newID] = spec;
+    s_lastRuns[newID] = block.timestamp;
+    s_handlerSignatures[newID] = handlerSig(target, handler);
+    s_nextCronJobID++;
+    emit CronJobCreated(newID, target, handler);
+  }
+
+  /**
+   * @notice validates the input to performUpkeep
+   * @param id the id of the cron job
+   * @param tickTime the observed tick time
+   * @param target the contract to forward the tx to
+   * @param handler the handler of the conract receiving the forwarded tx
+   */
+  function validate(
+    uint256 id,
+    uint256 tickTime,
+    address target,
+    bytes memory handler
+  ) private {
+    if (block.timestamp < tickTime) {
+      revert TickInFuture();
+    }
+    if (tickTime <= s_lastRuns[id]) {
+      revert TickTooOld();
+    }
+    if (!s_specs[id].matches(tickTime)) {
+      revert TickDoesntMatchSpec();
+    }
+    if (handlerSig(target, handler) != s_handlerSignatures[id]) {
+      revert InvalidHandler();
+    }
+  }
+
+  /**
+   * @notice returns a unique identifier for target/handler pairs
+   * @param target the contract to forward the tx to
+   * @param handler the handler of the conract receiving the forwarded tx
+   * @return a hash of the inputs
+   */
+  function handlerSig(address target, bytes memory handler)
+    private
+    pure
+    returns (bytes32)
+  {
+    return keccak256(abi.encodePacked(target, handler));
+  }
+
+  /**
+   * @notice delegates a function call to s_delegate, returning directly to the caller
+   */
+  function delegate() private {
```

Comment:
Any reason not to use the [OpenZeppelin implementation](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/Address.sol#L179) instead?

Reply:
this is c/p directly from OZ. I just omitted the rest of the contract since it felt it added unnecessary bulk. Happy to just use OZ if that feels cleaner. I guess the optimizer will probably trim most of the unused bulk...

Reply:
okay yeah let's just use OZ. added it back in.